File: agents/go/exporters.go
package openlog

import (
	"context"
	"net/url"
	"strings"
	"time"

	"go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploggrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	sdklog "go.opentelemetry.io/otel/sdk/log"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"google.golang.org/grpc/encoding/gzip"
)

type exporters struct {
	trace  sdktrace.SpanExporter
	metric sdkmetric.Exporter
	log    sdklog.Exporter
}

// newExporters builds the three OTLP exporters. Retries (429/502/503/504 over HTTP,
// UNAVAILABLE/RESOURCE_EXHAUSTED over gRPC) back off exponentially and honor
// Retry-After / RetryInfo from ingest (D-014).
func newExporters(ctx context.Context, cfg *Config) (*exporters, error) {
	u, _ := url.Parse(cfg.Endpoint)
	insecure := u.Scheme == "http"
	base := strings.TrimRight(u.Path, "/")
	r := cfg.retry

	if cfg.Protocol == ProtocolGRPC {
		hostport := u.Host
		var comp string
		if cfg.Compression == "gzip" {
			comp = gzip.Name
		}
		te, err := otlptrace.New(ctx, otlptracegrpc.NewClient(grpcTraceOpts(cfg, hostport, insecure, comp)...))
		if err != nil {
			return nil, err
		}
		mopts := []otlpmetricgrpc.Option{otlpmetricgrpc.WithEndpoint(hostport), otlpmetricgrpc.WithHeaders(cfg.Headers),
			otlpmetricgrpc.WithTimeout(cfg.ExportTimeout),
			otlpmetricgrpc.WithRetry(otlpmetricgrpc.RetryConfig{Enabled: true, InitialInterval: r.initial, MaxInterval: r.max, MaxElapsedTime: r.elapsed})}
		lopts := []otlploggrpc.Option{otlploggrpc.WithEndpoint(hostport), otlploggrpc.WithHeaders(cfg.Headers),
			otlploggrpc.WithTimeout(cfg.ExportTimeout),
			otlploggrpc.WithRetry(otlploggrpc.RetryConfig{Enabled: true, InitialInterval: r.initial, MaxInterval: r.max, MaxElapsedTime: r.elapsed})}
		if insecure {
			mopts = append(mopts, otlpmetricgrpc.WithInsecure())
			lopts = append(lopts, otlploggrpc.WithInsecure())
		}
		if comp != "" {
			mopts = append(mopts, otlpmetricgrpc.WithCompressor(comp))
			lopts = append(lopts, otlploggrpc.WithCompressor(comp))
		}
		me, err := otlpmetricgrpc.New(ctx, mopts...)
		if err != nil {
			return nil, err
		}
		le, err := otlploggrpc.New(ctx, lopts...)
		if err != nil {
			return nil, err
		}
		return &exporters{trace: te, metric: me, log: le}, nil
	}

	topts := []otlptracehttp.Option{otlptracehttp.WithEndpoint(u.Host), otlptracehttp.WithURLPath(base + "/v1/traces"),
		otlptracehttp.WithHeaders(cfg.Headers), otlptracehttp.WithTimeout(cfg.ExportTimeout),
		otlptracehttp.WithRetry(otlptracehttp.RetryConfig{Enabled: true, InitialInterval: r.initial, MaxInterval: r.max, MaxElapsedTime: r.elapsed})}
	mopts := []otlpmetrichttp.Option{otlpmetrichttp.WithEndpoint(u.Host), otlpmetrichttp.WithURLPath(base + "/v1/metrics"),
		otlpmetrichttp.WithHeaders(cfg.Headers), otlpmetrichttp.WithTimeout(cfg.ExportTimeout),
		otlpmetrichttp.WithRetry(otlpmetrichttp.RetryConfig{Enabled: true, InitialInterval: r.initial, MaxInterval: r.max, MaxElapsedTime: r.elapsed})}
	lopts := []otlploghttp.Option{otlploghttp.WithEndpoint(u.Host), otlploghttp.WithURLPath(base + "/v1/logs"),
		otlploghttp.WithHeaders(cfg.Headers), otlploghttp.WithTimeout(cfg.ExportTimeout),
		otlploghttp.WithRetry(otlploghttp.RetryConfig{Enabled: true, InitialInterval: r.initial, MaxInterval: r.max, MaxElapsedTime: r.elapsed})}
	if insecure {
		topts = append(topts, otlptracehttp.WithInsecure())
		mopts = append(mopts, otlpmetrichttp.WithInsecure())
		lopts = append(lopts, otlploghttp.WithInsecure())
	}
	if cfg.Compression == "gzip" {
		topts = append(topts, otlptracehttp.WithCompression(otlptracehttp.GzipCompression))
		mopts = append(mopts, otlpmetrichttp.WithCompression(otlpmetrichttp.GzipCompression))
		lopts = append(lopts, otlploghttp.WithCompression(otlploghttp.GzipCompression))
	}
	te, err := otlptracehttp.New(ctx, topts...)
	if err != nil {
		return nil, err
	}
	me, err := otlpmetrichttp.New(ctx, mopts...)
	if err != nil {
		return nil, err
	}
	le, err := otlploghttp.New(ctx, lopts...)
	if err != nil {
		return nil, err
	}
	return &exporters{trace: te, metric: me, log: le}, nil
}

func grpcTraceOpts(cfg *Config, hostport string, insecure bool, comp string) []otlptracegrpc.Option {
	r := cfg.retry
	o := []otlptracegrpc.Option{otlptracegrpc.WithEndpoint(hostport), otlptracegrpc.WithHeaders(cfg.Headers),
		otlptracegrpc.WithTimeout(cfg.ExportTimeout), otlptracegrpc.WithReconnectionPeriod(5 * time.Second),
		otlptracegrpc.WithRetry(otlptracegrpc.RetryConfig{Enabled: true, InitialInterval: r.initial, MaxInterval: r.max, MaxElapsedTime: r.elapsed})}
	if insecure {
		o = append(o, otlptracegrpc.WithInsecure())
	}
	if comp != "" {
		o = append(o, otlptracegrpc.WithCompressor(comp))
	}
	return o
}

File: agents/node/src/exporters.ts
import { OTLPLogExporter } from '@opentelemetry/exporter-logs-otlp-proto';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-proto';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-proto';
import type { LogRecordExporter } from '@opentelemetry/sdk-logs';
import type { PushMetricExporter } from '@opentelemetry/sdk-metrics';
import type { SpanExporter } from '@opentelemetry/sdk-trace-base';
import { createRequire } from 'node:module';
import * as path from 'node:path';
import { ConfigError, PROTOCOL_GRPC, type Config } from './config';

export interface Exporters {
  trace: SpanExporter;
  metric: PushMetricExporter;
  log: LogRecordExporter;
}

type Ctor<T> = new (config: Record<string, unknown>) => T;

function loadGrpc(pkg: string): Record<string, unknown> {
  // grpc exporters are optional peer dependencies: resolved from this package, then from the application.
  const bases = [__filename, path.join(process.cwd(), 'index.js')];
  let lastErr: unknown;
  for (const base of bases) {
    try {
      return createRequire(base)(pkg) as Record<string, unknown>;
    } catch (err) {
      lastErr = err;
    }
  }
  throw new ConfigError(
    `openlog: OPENLOG_PROTOCOL=grpc needs ${pkg} (npm install @opentelemetry/exporter-trace-otlp-grpc@0.222.0 ` +
      `@opentelemetry/exporter-metrics-otlp-grpc@0.222.0 @opentelemetry/exporter-logs-otlp-grpc@0.222.0): ${String(lastErr)}`,
  );
}

/**
 * OTLP exporters. http/protobuf posts to <endpoint>/v1/{traces,metrics,logs}; grpc uses the endpoint as target
 * (http:// = plaintext). The OTel exporters retry 429/502/503/504 (and gRPC UNAVAILABLE/RESOURCE_EXHAUSTED) with
 * exponential backoff and honor Retry-After (D-014); failures reach the rate-limited diagnostics logger.
 */
export function createExporters(cfg: Config): Exporters {
  const common = { headers: { ...cfg.headers }, compression: cfg.compression, timeoutMillis: cfg.exportTimeoutMs };
  if (cfg.protocol === PROTOCOL_GRPC) {
    const traceMod = loadGrpc('@opentelemetry/exporter-trace-otlp-grpc');
    const metricMod = loadGrpc('@opentelemetry/exporter-metrics-otlp-grpc');
    const logMod = loadGrpc('@opentelemetry/exporter-logs-otlp-grpc');
    const grpc = createRequire(require.resolve('@opentelemetry/exporter-trace-otlp-grpc', { paths: [__dirname, process.cwd()] }))(
      '@grpc/grpc-js',
    ) as { Metadata: new () => { set(k: string, v: string): void } };
    const metadata = (): unknown => {
      const md = new grpc.Metadata();
      for (const [k, v] of Object.entries(cfg.headers)) md.set(k, v);
      return md;
    };
    const g = { url: cfg.endpoint, compression: cfg.compression, timeoutMillis: cfg.exportTimeoutMs };
    return {
      trace: new (traceMod.OTLPTraceExporter as Ctor<SpanExporter>)({ ...g, metadata: metadata() }),
      metric: new (metricMod.OTLPMetricExporter as Ctor<PushMetricExporter>)({ ...g, metadata: metadata() }),
      log: new (logMod.OTLPLogExporter as Ctor<LogRecordExporter>)({ ...g, metadata: metadata() }),
    };
  }
  return {
    trace: new OTLPTraceExporter({ ...common, url: `${cfg.endpoint}/v1/traces` } as ConstructorParameters<typeof OTLPTraceExporter>[0]),
    metric: new OTLPMetricExporter({ ...common, url: `${cfg.endpoint}/v1/metrics` } as ConstructorParameters<typeof OTLPMetricExporter>[0]),
    log: new OTLPLogExporter({ ...common, url: `${cfg.endpoint}/v1/logs` } as ConstructorParameters<typeof OTLPLogExporter>[0]),
  };
}
